### PR TITLE
feat: add hypertune for feature flags

### PR DESCRIPTION
**Pull request by Sparkier**
Depends on https://github.com/end-of-text/endoftext/pull/80

### CODE REVIEW

**Comment by cabreraalex** [comment]
todo: rebase with new changes and make first feature flag implementation to be able to turn on/off editors.

**Comment by Sparkier** [comment]
This adds a flag for each editor. There is some question about behavior. I implemented it so that we check on generation of suggestions, but once suggestions are generated there is no check of the feature flag anymore (e.g. people can accept a suggestion even if they could not generate that very suggestion because of a feature flag). I think this makes sene but it's not the only option.

File: src/lib/hypertune/hypertune.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { initializeHypertune } from './generated';
+
+const hypertune = initializeHypertune({}, { token: PUBLIC_HYPERTUNE_TOKEN });
+
+export async function getRootNode(user: { id: string; email?: string }) {
+	await hypertune.initFromServerIfNeeded();
+
+	return hypertune.root({
+		context: {
+			environment: 'DEVELOPMENT',
```
Comment (cabreraalex) [inline src/lib/hypertune/hypertune.ts:11]: is this supposed to be development

File: src/lib/hypertune/hypertune.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { PUBLIC_HYPERTUNE_TOKEN } from '$env/static/public';
+import { initializeHypertune } from './generated';
+
+const hypertune = initializeHypertune({}, { token: PUBLIC_HYPERTUNE_TOKEN });
+
+export async function getRootNode(user: { id: string; email?: string }) {
```
Comment (cabreraalex) [inline src/lib/hypertune/hypertune.ts:6]: better name? getRootNode very generic

File: src/lib/server/editors/editors.ts [hunk 3]
```diff
@@ -1,23 +1,51 @@
...
-	new ShortenEditor(),
-	new LongerEditor(),
-	new JSONInstructionEditor(),
-	new JSONDescriptionEditor(),
-	new OutputDescriptionEditor(),
-	new CompressionEditor(),
-	new ConvertToJSONEditor(),
-	new PersonaEditor()
-];
+export function getEditors(hypertuneRoot?: RootNode): PromptEditor[] {
```
Comment (cabreraalex) [inline src/lib/server/editors/editors.ts:14]: Why optional?

File: src/routes/(app)/project/[id]/[promptId]/+page.server.ts [hunk 4]
```diff
@@ -55,7 +61,8 @@ export async function load({ locals: { supabase }, params }) {
 		instancesReq.data || [],
 		promptsRes.data[0]
 	);
-	const suggestions = getSuggestions(supabase, promptsRes.data[0], false);
+	const hypertuneRoot = await getRootNode(session.user);
```
Comment (cabreraalex) [inline src/routes/(app)/project/[id]/[promptId]/+page.server.ts:64]: This doesn't change for every request right? Could we set this as a global or something like supabase? expensive to call every time.

**Review by cabreraalex — CHANGES_REQUESTED** [review]

File: src/routes/(app)/project/[id]/[promptId]/+page.server.ts (hunk 4 above)
Comment (cabreraalex) [reply src/routes/(app)/project/[id]/[promptId]/+page.server.ts:64]: jk it's per user and it's a REST endpoint. maybe too complicated to cache somehow

File: src/lib/server/editors/editors.ts (hunk 3 above)
Comment (Sparkier) [reply src/lib/server/editors/editors.ts:14]: Because we don't filter on apply.

File: src/lib/hypertune/hypertune.ts (hunk 1 above)
Comment (Sparkier) [reply src/lib/hypertune/hypertune.ts]: Doesn't matter you can activate/deactivate them based on the env but we don't currently so I don't think it matters.

**Review by cabreraalex — APPROVED** [review]

**Pull request merged**